hail/python/hail/experimental/dnd/array.py

```diff
@@ -0,0 +1,260 @@
+
+        assert 'y' not in mt.row
+        assert 'x' not in mt.row
+        assert 'block' not in mt.row
+
+        n_rows, n_cols = mt.count()
+        n_block_rows = (n_rows + block_size - 1) // block_size
+        n_block_cols = (n_cols + block_size - 1) // block_size
+        entries, cols, row_index, col_blocks = (Env.get_uid() for _ in range(4))
+        mt = (mt
+              .select_globals()
+              .select_rows()
+              .select_cols()
+              .add_row_index(row_index)
+              .localize_entries(entries, cols))
+        # FIXME: remove when ndarray support structs
+        mt = mt.annotate(**{entries: mt[entries][entry_field]})
+        mt = mt.annotate(
+            **{col_blocks: hl.range(n_block_cols).map(
+                lambda y: hl.struct(
+                    y=y,
+                    entries=mt[entries][(y * block_size):((y + 1) * block_size)]))}
+        )
+        mt = mt.explode(col_blocks)
+        mt = mt.select(row_index, **mt[col_blocks])
+        mt = mt.annotate(x=hl.int(mt[row_index] // block_size))
+        mt = mt.key_by(mt.x, mt.y)
+        mt = mt.group_by(mt.x, mt.y).aggregate(
+            entries=hl.sorted(
+                hl.agg.collect(hl.struct(row_index=mt[row_index], entries=mt.entries)),
+                key=lambda x: x.row_index
+            ).map(lambda x: x.entries))
+        mt = mt.select(block=hl.nd.array(mt.entries))
+        mt = mt.select_globals(
+            x_field='x',
+            y_field='y',
+            n_rows=n_rows,
+            n_cols=n_cols,
+            n_block_rows=n_block_rows,
+            n_block_cols=n_block_cols,
+            block_size=block_size)
+        fname = new_temp_file()
+        mt = mt.key_by(mt.x, mt.y)
+        mt.write(fname, _codec_spec=DNDArray.fast_codec_spec)
+        t = hl.read_table(fname, _intervals=[
+            hl.Interval(hl.Struct(x=x, y=y),
+                        hl.Struct(x=x, y=y + 1))
+            for x in range(n_block_rows)
+            for y in range(n_block_cols)])
+        return DNDArray(t)
+
+    def __init__(self, t: Table) -> 'DNDArray':
+        assert 'x' in t.row
+        assert 'y' in t.row
+        assert 'block' in t.row
+
+        self.m: Table = t
+
+        dimensions = t.globals.collect()[0]
+        self.x_field: str = dimensions.x_field
+        self.y_field: str = dimensions.y_field
+        self.n_rows: int = dimensions.n_rows
+        self.n_cols: int = dimensions.n_cols
+        self.n_block_rows: int = dimensions.n_block_rows
+        self.n_block_cols: int = dimensions.n_block_cols
+        self.block_size: int = dimensions.block_size
+
+        assert self.n_block_rows == (self.n_rows + self.block_size - 1) // self.block_size
+        assert self.n_block_cols == (self.n_cols + self.block_size - 1) // self.block_size
+
+    def count_blocks(self) -> Tuple[int, int]:
+        return (self.n_block_cols, self.n_block_rows)
+
+    @property
+    def T(self) -> 'DNDArray':
```

Comment:
In numpy and in `hail.ndarray`, both `T` and `transpose()` are valid. We should support both here too I think?